File: 10/gemini-3.0-pro/typescript/solution.ts
import * as fs from 'fs';

function solve() {
    const input = fs.readFileSync('input.txt', 'utf-8');
    let totalPressesPart1 = 0;
    let totalPressesPart2 = 0;

    const lines = input.split('\n');
    for (const line of lines) {
        if (!line.trim()) continue;
        totalPressesPart1 += solvePart1(line);
        totalPressesPart2 += solvePart2(line);
    }

    console.log(`Day 10 Part 1: ${totalPressesPart1}`);
    console.log(`Day 10 Part 2: ${totalPressesPart2}`);
}

function solvePart1(line: string): number {
    const targetMatch = line.match(/\[([.#]+)\]/);
    if (!targetMatch) return 0;
    const targetStr = targetMatch[1];
    const L = targetStr.length;
    const target = new Array(L).fill(0);
    for (let i = 0; i < L; i++) {
        target[i] = targetStr[i] === '#' ? 1 : 0;
    }

    const buttonMatches = Array.from(line.matchAll(/\(([\d,]+)\)/g));
    const buttons: number[][] = [];
    for (const bm of buttonMatches) {
        const b = new Array(L).fill(0);
        const parts = bm[1].split(',');
        for (const p of parts) {
            const idx = parseInt(p);
            if (!isNaN(idx) && idx < L) {
                b[idx] = 1;
            }
        }
        buttons.push(b);
    }

    const B = buttons.length;
    const M: number[][] = [];
    for (let r = 0; r < L; r++) {
        M[r] = new Array(B + 1).fill(0);
        for (let c = 0; c < B; c++) {
            M[r][c] = buttons[c][r];
        }
        M[r][B] = target[r];
    }

    let pivotRow = 0;
    const pivotCols: number[] = [];
    const colToPivotRow = new Array(B).fill(-1);

    for (let c = 0; c < B && pivotRow < L; c++) {
        let sel = -1;
        for (let r = pivotRow; r < L; r++) {
            if (M[r][c] === 1) {
                sel = r;
                break;
            }
        }

        if (sel === -1) continue;

        // Swap
        [M[pivotRow], M[sel]] = [M[sel], M[pivotRow]];

        // Eliminate
        for (let r = 0; r < L; r++) {
            if (r !== pivotRow && M[r][c] === 1) {
                for (let k = c; k <= B; k++) {
                    M[r][k] ^= M[pivotRow][k];
                }
            }
        }

        pivotCols.push(c);
        colToPivotRow[c] = pivotRow;
        pivotRow++;
    }

    for (let r = pivotRow; r < L; r++) {
        if (M[r][B] === 1) return 0;
    }

    const freeCols: number[] = [];
    for (let c = 0; c < B; c++) {
        if (!pivotCols.includes(c)) freeCols.push(c);
    }

    let minPresses = Number.MAX_SAFE_INTEGER;
    const numFree = freeCols.length;
    const combinations = 1 << numFree;

    for (let i = 0; i < combinations; i++) {
        const x = new Array(B).fill(0);
        let currentPresses = 0;

        for (let k = 0; k < numFree; k++) {
            if ((i & (1 << k)) !== 0) {
                x[freeCols[k]] = 1;
                currentPresses++;
            }
        }

        for (const pCol of pivotCols) {
            const r = colToPivotRow[pCol];
            let val = M[r][B];
            for (const fCol of freeCols) {
                if (M[r][fCol] === 1 && x[fCol] === 1) {
                    val ^= 1;
                }
            }
            x[pCol] = val;
            if (val === 1) currentPresses++;
        }

        if (currentPresses < minPresses) {
            minPresses = currentPresses;
        }
    }

    return minPresses;
}

function solvePart2(line: string): number {
    const targetMatch = line.match(/\{([\d,]+)\}/);
    if (!targetMatch) return 0;
    const targetParts = targetMatch[1].split(',').map(Number);
    const L = targetParts.length;
    const b = targetParts;

    const buttonMatches = Array.from(line.matchAll(/\(([\d,]+)\)/g));
    const A_cols: number[][] = [];
    for (const bm of buttonMatches) {
        const col = new Array(L).fill(0);
        const parts = bm[1].split(',');
        for (const p of parts) {
            const idx = parseInt(p);
            if (!isNaN(idx) && idx < L) {
                col[idx] = 1;
            }
        }
        A_cols.push(col);
    }

    const B = A_cols.length;
    const matrix: number[][] = [];
    for (let r = 0; r < L; r++) {
        matrix[r] = new Array(B + 1).fill(0);
        for (let c = 0; c < B; c++) {
            matrix[r][c] = A_cols[c][r];
        }
        matrix[r][B] = b[r];
    }

    let pivotRow = 0;
    const pivotCols: number[] = [];
    const colToPivotRow = new Array(B).fill(-1);

    for (let c = 0; c < B && pivotRow < L; c++) {
        let sel = -1;
        for (let r = pivotRow; r < L; r++) {
            if (Math.abs(matrix[r][c]) > 1e-9) {
                sel = r;
                break;
            }
        }

        if (sel === -1) continue;

        [matrix[pivotRow], matrix[sel]] = [matrix[sel], matrix[pivotRow]];

        const pivotVal = matrix[pivotRow][c];
        for (let k = c; k <= B; k++) matrix[pivotRow][k] /= pivotVal;

        for (let r = 0; r < L; r++) {
            if (r !== pivotRow && Math.abs(matrix[r][c]) > 1e-9) {
                const factor = matrix[r][c];
                for (let k = c; k <= B; k++) matrix[r][k] -= factor * matrix[pivotRow][k];
            }
        }

        pivotCols.push(c);
        colToPivotRow[c] = pivotRow;
        pivotRow++;
    }

    for (let r = pivotRow; r < L; r++) {
        if (Math.abs(matrix[r][B]) > 1e-9) return 0;
    }

    const freeCols: number[] = [];
    for (let c = 0; c < B; c++) {
        if (!pivotCols.includes(c)) freeCols.push(c);
    }

    const UBs = new Array(B).fill(0);
    for (let c = 0; c < B; c++) {
        let minUb = Number.MAX_SAFE_INTEGER;
        let bounded = false;
        for (let r = 0; r < L; r++) {
            if (A_cols[c][r] > 0.5) {
                const val = Math.floor(b[r] / A_cols[c][r]);
                if (val < minUb) minUb = val;
                bounded = true;
            }
        }
        UBs[c] = bounded ? minUb : 0;
    }

    let minTotalPresses = Number.MAX_SAFE_INTEGER;
    let found = false;

    function search(freeIdx: number, currentFreeVals: number[]) {
        if (freeIdx === freeCols.length) {
            let currentPresses = 0;
            for (const v of currentFreeVals) currentPresses += v;

            let possible = true;
            for (const pCol of pivotCols) {
                const r = colToPivotRow[pCol];
                let val = matrix[r][B];
                for (let j = 0; j < freeCols.length; j++) {
                    const fCol = freeCols[j];
                    val -= matrix[r][fCol] * currentFreeVals[j];
                }

                if (val < -1e-9) { possible = false; break; }
                const longVal = Math.round(val);
                if (Math.abs(val - longVal) > 1e-9) { possible = false; break; }
                currentPresses += longVal;
            }

            if (possible) {
                if (currentPresses < minTotalPresses) {
                    minTotalPresses = currentPresses;
                    found = true;
                }
            }
            return;
        }

        const fColIdx = freeCols[freeIdx];
        const limit = UBs[fColIdx];

        for (let val = 0; val <= limit; val++) {
            currentFreeVals[freeIdx] = val;
            search(freeIdx + 1, currentFreeVals);
        }
    }

    search(0, new Array(freeCols.length).fill(0));

    return found ? minTotalPresses : 0;
}

solve();

File: 10/gemini-3.0-pro/python/solution.py
import re
import sys
import math

sys.setrecursionlimit(2000)

def solve():
    try:
        with open('input.txt', 'r') as f:
            input_data = f.read()
    except FileNotFoundError:
        print("input.txt not found")
        return

    total_presses_part1 = 0
    total_presses_part2 = 0
    lines = input_data.split('\n')
    for line in lines:
        if not line.strip():
            continue
        total_presses_part1 += solve_machine(line)
        total_presses_part2 += solve_part2(line)

    print(f"Day 10 Part 1: {total_presses_part1}")
    print(f"Day 10 Part 2: {total_presses_part2}")

def solve_machine(line):
    target_match = re.search(r'\[([.#]+)\]', line)
    if not target_match:
        return 0
    target_str = target_match.group(1)
    L = len(target_str)
    target = [1 if c == '#' else 0 for c in target_str]

    button_matches = re.findall(r'\(([\d,]+)\)', line)
    buttons = []
    for bm in button_matches:
        b = [0] * L
        parts = bm.split(',')
        for p in parts:
            try:
                idx = int(p)
                if idx < L:
                    b[idx] = 1
            except ValueError:
                pass
        buttons.append(b)

    B = len(buttons)
    # Matrix M of size L x (B + 1)
    M = [[0] * (B + 1) for _ in range(L)]
    for r in range(L):
        for c in range(B):
            M[r][c] = buttons[c][r]
        M[r][B] = target[r]

    pivot_row = 0
    pivot_cols = []
    col_to_pivot_row = [-1] * B

    for c in range(B):
        if pivot_row >= L:
            break

        sel = -1
        for r in range(pivot_row, L):
            if M[r][c] == 1:
                sel = r
                break

        if sel == -1:
            continue

        # Swap
        M[pivot_row], M[sel] = M[sel], M[pivot_row]

        # Eliminate
        for r in range(L):
            if r != pivot_row and M[r][c] == 1:
                for k in range(c, B + 1):
                    M[r][k] ^= M[pivot_row][k]

        pivot_cols.append(c)
        col_to_pivot_row[c] = pivot_row
        pivot_row += 1

    for r in range(pivot_row, L):
        if M[r][B] == 1:
            return 0

    free_cols = [c for c in range(B) if c not in pivot_cols]

    min_presses = float('inf')
    num_free = len(free_cols)
    combinations = 1 << num_free

    for i in range(combinations):
        x = [0] * B
        current_presses = 0

        for k in range(num_free):
            if (i & (1 << k)) != 0:
                x[free_cols[k]] = 1
                current_presses += 1

        for p_col in pivot_cols:
            r = col_to_pivot_row[p_col]
            val = M[r][B]
            for f_col in free_cols:
                if M[r][f_col] == 1 and x[f_col] == 1:
                    val ^= 1
            x[p_col] = val
            if val == 1:
                current_presses += 1

        if current_presses < min_presses:
            min_presses = current_presses

    return min_presses

def solve_part2(line):
    target_match = re.search(r'\{([\d,]+)\}', line)
    if not target_match:
        return 0
    target_parts = [int(x) for x in target_match.group(1).split(',')]
    L = len(target_parts)
    b_vec = target_parts

    button_matches = re.findall(r'\(([\d,]+)\)', line)
    A_cols = []
    for bm in button_matches:
        col = [0] * L
        parts = bm.split(',')
        for p in parts:
            try:
                idx = int(p)
                if idx < L:
                    col[idx] = 1
            except ValueError:
                pass
        A_cols.append(col)

    B = len(A_cols)
    matrix = [[0.0] * (B + 1) for _ in range(L)]
    for r in range(L):
        for c in range(B):
            matrix[r][c] = float(A_cols[c][r])
        matrix[r][B] = float(b_vec[r])

    pivot_row = 0
    pivot_cols = []
    col_to_pivot_row = [-1] * B

    for c in range(B):
        if pivot_row >= L:
            break

        sel = -1
        for r in range(pivot_row, L):
            if abs(matrix[r][c]) > 1e-9:
                sel = r
                break

        if sel == -1:
            continue

        matrix[pivot_row], matrix[sel] = matrix[sel], matrix[pivot_row]

        pivot_val = matrix[pivot_row][c]
        for k in range(c, B + 1):
            matrix[pivot_row][k] /= pivot_val

        for r in range(L):
            if r != pivot_row and abs(matrix[r][c]) > 1e-9:
                factor = matrix[r][c]
                for k in range(c, B + 1):
                    matrix[r][k] -= factor * matrix[pivot_row][k]

        pivot_cols.append(c)
        col_to_pivot_row[c] = pivot_row
        pivot_row += 1

    for r in range(pivot_row, L):
        if abs(matrix[r][B]) > 1e-9:
            return 0

    free_cols = [c for c in range(B) if c not in pivot_cols]

    UBs = [0] * B
    for c in range(B):
        min_ub = float('inf')
        bounded = False
        for r in range(L):
            if A_cols[c][r] > 0.5:
                val = math.floor(b_vec[r] / A_cols[c][r])
                if val < min_ub:
                    min_ub = val
                bounded = True
        UBs[c] = min_ub if bounded else 0

    min_total_presses = float('inf')
    found = False

    def search(free_idx, current_free_vals):
        nonlocal min_total_presses, found
        if free_idx == len(free_cols):
            current_presses = sum(current_free_vals)
            possible = True

            for p_col in pivot_cols:
                r = col_to_pivot_row[p_col]
                val = matrix[r][B]
                for j, f_col in enumerate(free_cols):
                    val -= matrix[r][f_col] * current_free_vals[j]

                if val < -1e-9:
                    possible = False
                    break
                long_val = round(val)
                if abs(val - long_val) > 1e-9:
                    possible = False
                    break
                current_presses += long_val

            if possible:
                if current_presses < min_total_presses:
                    min_total_presses = current_presses
                    found = True
            return

        f_col_idx = free_cols[free_idx]
        limit = UBs[f_col_idx]

        for val in range(limit + 1):
            current_free_vals[free_idx] = val
            search(free_idx + 1, current_free_vals)
            current_free_vals[free_idx] = 0 # Backtrack

    search(0, [0] * len(free_cols))

    return int(min_total_presses) if found else 0

if __name__ == '__main__':
    solve()
